docs/proposal-subscriptions.md

```diff
@@ -0,0 +1,280 @@
+        source
+        target
+      }
+    }
+  }
+}
+```
+
+These subscriptions are different but they have a large overlap. If we were to register both subscriptions with the UIS we would end up with two similar but different data stores in the UI. This would be in-efficient and opens up discontinuity issues.
+
+Users may want to open many similar but different views on the same workflow so we need to be able to handle this in an elegant manner.
+
+#### Merging Subscriptions (2)
+
+One solution to this problem is to "merge" subscriptions together.
+
+Subscriptions are effectively "sets" of data fields, two subscriptions can be "merged" by taking their union. For example the union of the previous two subscriptions is this:
+
+```
+subscription {
+  workflows(ids: [<id>]) {
+    taskProxies {
+      name
+      id
+      status
+      firstParent {
+        id
+      }
+    }
+    familyProxies {
+      id
+      name
+      firstParent {
+      }
+    }
+    nodesEdges {
+      nodes {
+        id
+      }
+      edges {
+        source
+        target
+      }
+    }
+  }
+}
+```
+
+As an added bonus we can query the UI to find out exactly what it is subscribing to and exactly which view has asked for what data making optimisation/debugging much nicer.
+
+### The Workflow Server
+
+When events occur the WS sends "deltas" containing only the changed information.
+
+For example if a job has changed state from submitted to running we might expect a delta of the form:
+
+```
+taskProxies {
+  id: <id>,
+  status: "running"
+  jobs {
+    id: <id>,
+    status: "running"
+  }
+}
+```
+
+### The UI Server
+
+#### Subscriptions (3)
+
+When subscriptions are registered with the UIS they are stored locally in a mapping so we can associate subscriptions with open websockets.
+
+#### Deltas (4)
+
+When deltas arrive from the WS we loop through the subscriptions and take the intersection of the delta and subscription fields, if there is an overlap, this is the delta we send to the UI.
+
+#### Overlaps (5)
+
+For example using the previous delta and subscription:
+
+
+<table>
+  <th>
+    <td>Delta</td>
+    <td>Subscription</td>
+    <td>Intersection</td>
+  </th>
+  <tr>
+    <td>
+        Example 1
+    </td>
+    <td>
+      <pre>
+workflows(ids: [<id>]) {
+  taskProxies {
+    id: <id>,
+    status: "running"
+    jobs {
+      id: <id>,
+      status: "running"
+    }
+  }
+}
+      </pre>
+    </td>
+    <td>
+      <pre>
+subscription {
+  workflows(ids: [<id>]) {
+    taskProxies {
+      name
+      id
+      status
+      firstParent {
+        id
+      }
+    }
+    familyProxies {
+      id
+      name
+      firstParent {
+      }
+    }
+    nodesEdges {
+      nodes {
+        id
+      }
+      edges {
+        source
+        target
+      }
+    }
+  }
+}
+      </pre>
+    </td>
+    <td>
+      <pre>
+workflows(ids: [<id>]) {
+  taskProxies {
+    id: <id>,
+    status: "running"
+  }
+}
+      </pre>
+    </td>
+  </tr>
+</table>
```

Comment:
Because deltas are actually flat (only one topic at a time, i.e. `workflow` OR `taskProxy`), then we would need to drive the subscription off the data-store and not filter out the fields parent to the delta (even though the parent is not in the delta).. Otherwise we would need to enforce flat subscriptions, and probably construct new delta graphql object types to send the context of the delta (like protobuf deltas).

Consider:
```
workflows(ids: []) {
  taskProxies {
    id: ,
    status: "running"
    jobs {
      id: ,
      status: "running"
    }
  }
}
```
If only `jobs` topic is sent by the WFS, then the delta may have no information about `taskProxy` or `workflow`.. The delta itself is received at the UIS with the workflow and topic context though...

Reply:
Is this not what the intersection idea solves?

Using your example:

*Subscription:*

```graphql
workflows(ids: ["u|w"]) {
  taskProxies {
    id
    status
    jobs {
      id
      status
      submitNumber
      batchSystem
    }
  }
}
```

*Delta:*

```json
{
  "workflows": {
    "id": "u|w",
    "taskProxies": [
      {
        "id": "u|w|c|t",
        "jobs": {
          "id": "u|w|c|t|j",
          "status": "running"
        }
      }
    ]
  }
}
```

*Intersection:*

```json
{
  "workflows": {
    "id": "u|w",
    "taskProxies": [
      {
        "id": "u|w|c|t",
        "jobs": {
          "id": "u|w|c|t|j",
          "status": "running"
        }
      }
    ]
  }
}
```

Reply:
My point is; if your delta is flat and all you receive at the UIS is:
```
        "jobs": {
          "id": "u|w|c|t|j",
          "status": "running"
        }
```
Then you don't/may-not have the information in the delta to resolve the other bits:
```
{
  "workflows": {
    "id": "u|w",
    "taskProxies": [
      {
        "id": "u|w|c|t",
.
.
.
```
(meaning you get nothing, not even jobs)

So you will need to use the data-store to fill in the missing bits before sending it on to the WUI.. This isn't hard to do, and we could make it so if the WUI subscribes to parent bits that aren't in the driving deltas then it (optionally) only receives the skeleton (IDs) of the parent objects..

The branches I'm working on clump related deltas together (which will help).. anyway these are finer points to resolve along the way.

Reply:
> So you will need to use the data-store to fill in the missing bits before sending it on to the WUI.

Not for subscriptions because the required data will already have been synced, should be no need to go to the data store.

The delta between the Scheduler and UIS should be exactly the same as the delta between the UIS and WUI.

Reply:
> Not for subscriptions because the required data will already have been synced

which means your subscription should be flat,  which is my contention.

Reply:
> Not for subscriptions because the required data will already have been synced

which means your subscription should be flat,  which is my contention.